test/exclude-members.test.js

```diff
@@ -607,3 +607,83 @@ test('Can exclude a person in a group with a lot of members', async (t) => {
   await p(alice.close)(true)
   await Promise.all(peers.map((peer) => p(peer.close)(true)))
 })
+
+test("restarting the client doesn't make us rejoin old stuff", async (t) => {
+  const alice = Testbot({
+    keys: ssbKeys.generate(null, 'alice'),
+    mfSeed: Buffer.from(
+      '000000000000000000000000000000000000000000000000000000000000a1ce',
+      'hex'
+    ),
+  })
+  let bob = Testbot({
+    name: 'bobrestart',
+    keys: ssbKeys.generate(null, 'bob'),
+    mfSeed: Buffer.from(
+      '0000000000000000000000000000000000000000000000000000000000000b0b',
+      'hex'
+    ),
+  })
+
+  await Promise.all([alice.tribes2.start(), bob.tribes2.start()])
+
+  const bobRoot = await p(bob.metafeeds.findOrCreate)()
+
+  await replicate(alice, bob).catch(t.error)
+
+  const { id: groupId } = await alice.tribes2
+    .create()
+    .catch((err) => t.error(err, 'alice failed to create group'))
+
+  await alice.tribes2
+    .addMembers(groupId, [bobRoot.id])
+    .then(() => t.pass('added bob'))
+    .catch((err) => t.error(err, 'add bob fail'))
+
+  await replicate(alice, bob).catch(t.error)
+
+  await bob.tribes2.acceptInvite(groupId)
+
+  await replicate(alice, bob).catch(t.error)
+
+  await alice.tribes2
+    .excludeMembers(groupId, [bobRoot.id])
+    .then(() => t.pass('alice excluded bob'))
+    .catch((err) => t.error(err, 'remove member fail'))
+
+  await replicate(alice, bob).catch(t.error)
+
+  t.deepEquals(
+    await bob.tribes2.get(groupId),
+    { id: groupId, excluded: true },
+    "bob knows he's excluded from the group before restart"
+  )
+
+  await p(bob.close)(true).then(() => t.pass("bob's client was closed"))
+  bob = Testbot({
+    rimraf: false,
+    name: 'bobrestart',
+    keys: ssbKeys.generate(null, 'bob'),
+    mfSeed: Buffer.from(
+      '0000000000000000000000000000000000000000000000000000000000000b0b',
+      'hex'
+    ),
+  })
+  t.pass('bob got a new client')
+  await bob.tribes2.start().then(() => t.pass('bob restarted'))
```

Comment:
```
ok 5 bob got a new client
ssb-keyring info-encoding found envelope-id-based-metafeed-announced-curve25519 string
ssb-keyring info-encoding found envelope-id-based-metafeed-announced-curve25519 string
/home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/envelope-js@1.3.2/node_modules/envelope-js/util/key-slot.js:27
    if (!recipient.key) throw new Error('envelope/key-slot: recipient is expected to have a key')
                   ^

TypeError: Cannot read properties of undefined (reading 'key')
    at flip (/home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/envelope-js@1.3.2/node_modules/envelope-js/util/key-slot.js:27:20)
    at setFlip (/home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/envelope-js@1.3.2/node_modules/envelope-js/util/key-slot.js:16:13)
    at unboxKey (/home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/envelope-js@1.3.2/node_modules/envelope-js/unbox.js:45:5)
    at unbox (/home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/envelope-js@1.3.2/node_modules/envelope-js/unbox.js:18:20)
    at Object.decrypt (/home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/ssb-box2@6.2.0/node_modules/ssb-box2/format.js:306:25)
    at decryptAndReconstruct (/home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/ssb-db2@6.3.3/node_modules/ssb-db2/indexes/private.js:146:43)
    at Object.decrypt (/home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/ssb-db2@6.3.3/node_modules/ssb-db2/indexes/private.js:213:31)
    at Object.o.write (/home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/ssb-db2@6.3.3/node_modules/ssb-db2/log.js:179:40)
    at Stream._writeToSink (/home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/async-append-only-log@4.3.10/node_modules/async-append-only-log/stream.js:85:30)
    at Stream._handleBlock (/home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/async-append-only-log@4.3.10/node_modules/async-append-only-log/stream.js:122:12)
```

Getting this, on start() it seems. If I remove the exclusion then this error doesn't appear, although the weird keyring logs at the start still do.

Reply:
Added some more logs

```
ok 5 bob got a new client
running start for @3XVgQP3VFF4r+YMtJk3QgOSz5zAWvfZXS0zYfqppf14=.ed25519
about to find additions feed
about to find root feed
  ssb:db2:base got index status: { version: 2, offset: 20794, processed: 24 } +0ms
  ssb:db2:keys got index status: { version: 1, offset: 20794, processed: 24 } +0ms
  ssb:db2:ebt got index status: { version: 1, offset: 20794, processed: 24 } +0ms
ssb-keyring info-encoding found envelope-id-based-metafeed-announced-curve25519 string
ssb-keyring info-encoding found envelope-id-based-metafeed-announced-curve25519 string
  ssb:db2:private decrypted loaded, size: 6 +0ms
  ssb:db2:private encrypted-box2 loaded, size: 2 +0ms
  ssb:db2:private loaded offset 18333 +1ms
  ssb:db2 lowest offset for all indexes is 18333 +0ms
/home/me/prj/ssb/ssb-tribes2/node_modules/.pnpm/envelope-js@1.3.2/node_modules/envelope-js/util/key-slot.js:27
    if (!recipient.key) throw new Error('envelope/key-slot: recipient is expected to have a key')
                   ^
...
```

I'm getting a hunch that messages that we previously were able to decrypt are now marked as decryptable, but since we've thrown out the old keys (we were excluded) we don't manage to decrypt anymore. Would we need to reset the index or something?

Reply:
tried adding a reindexEncrypted after finding and applying an exclusion and a timeout in the test but doesn't seem to help

Reply:
had a call with andre. he didn't think it made sense to remove the keys to a group (while keeping the messages). so i'm gonna call deleteFeed https://github.com/ssbc/ssb-db2#deletefeedfeedid-cb on all the feeds in a group when we're in the process of excluding ourselves from it

Reply:
https://github.com/ssbc/ssb-tribes2/issues/102 and https://github.com/ssbc/ssb-tribes2/issues/103

Reply:
Seems to be fixed by https://github.com/ssbc/ssb-box2/pull/32